### ISSUE TITLE
Source forest creation failed - on re-builds

### ISSUE THREAD

**Issue by svendjo** [google/bazel#291]
I'm following the steps in the getting started guide. And I've run into a problem. Re-building consistently fails.

The build succeeds again if I bazel clean the workspace, but that's not a viable solution for large projects..

skjohann@svendjohannsen-OptiPlex-9020:~/workspace/ws0709-bazel2$ bazel clean
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
skjohann@svendjohannsen-OptiPlex-9020:~/workspace/ws0709-bazel2$ bazel build bazel-example:my-other-runner
INFO: Found 1 target...
Target //bazel-example:my-other-runner up-to-date:
  bazel-bin/bazel-example/my-other-runner.jar
  bazel-bin/bazel-example/my-other-runner
INFO: Elapsed time: 1.330s, Critical Path: 0.78s
skjohann@svendjohannsen-OptiPlex-9020:~/workspace/ws0709-bazel2$ bazel build bazel-example:my-other-runner
INFO: Found 1 target...
ERROR: Source forest creation failed: /home/skjohann/.cache/bazel/_bazel_skjohann/ba8eedfaca24bfaac913e5fd6e9ef5c8/ws0709-bazel2/bazel-example (File exists); build aborted: /home/skjohann/.cache/bazel/_bazel_skjohann/ba8eedfaca24bfaac913e5fd6e9ef5c8/ws0709-bazel2/bazel-example (File exists).
INFO: Elapsed time: 0.069s

**Comment by svendjo** [comment]
Somehow it matter where my bazel WORKSPACE file is in the tree.

/ws0709-bazel2/bazel-example/java/com/example
....^.................................................................^....
WORKSPACE...........................................BUILD
ws0709-bazel2$ bazel build bazel-example/java/com/example:my-runner (ERROR)

/ws0709-bazel2/bazel-example/java/com/example
..............................^.......................................^....
......................WORKSPACE.....................BUILD
bazel-example$ bazel build java/com/example:my-runner (OK)

In short it works just fine if I put my WORKSPACE in the same directory as my git project, but if I put my WORKSPACE one directory up from the git project it fails to re-build.

**Comment by kchodorow** [comment]
I don't know what the errors you were getting were, but the location of the WORKSPACE file defines how all of the packages are referenced.  So if you have it in /ws0709-bazel2, then packages will be relative to that (//bazel-examples/java/com/example:my-runner). If it's in bazel-example, the target would be //java/com/example:whatever.  This matters for any dependencies example:my-runner is referencing.

**Comment by svendjo** [comment]
Right, I use a different path to build depending on where the WORKSPACE file is. The BUILD file does not specify any dependencies. The errors are in the first post.

**Comment by svendjo** [comment]
The issue seems to be that I named the directory "bazel-example" -- bazel uses bazel-* directories for its working directories.

**Comment by kchodorow** [comment]
Oh, yeah, that would do it.  bazel- directories are treated differently in the symlink forest creation.  Can you use a different name?

**Comment by svendjo** [comment]
Yep. Thanks.

**Comment by ohler** [comment]
I'm running into similar errors with a directory called `_build`.  Are directory names starting with `_` also problematic?  Is the complete list of reserved names documented somewhere?

**Comment by kchodorow** [comment]
Directories starting with _ should be fine, what's the error you're getting?

**Comment by ohler** [comment]
The error is very similar to the OP's.  First build works, second fails with "Source forest creation failed (File exists)":

```
ohler$ bazel test backend/src/...
INFO: Found [...] targets and [...] test targets...
INFO: Elapsed time: [...]s, Critical Path: [...]s
[...]
Executed [...] out of [...] tests: [...] tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
ohler$ bazel test backend/src/...
INFO: Found [...] targets and [...] test targets...
ERROR: Source forest creation failed: /private/var/tmp/_bazel_ohler/58c6dc864860181b2aa138aee6d4ab43/dir/_build (File exists).
INFO: Elapsed time: 1.725s
ERROR: Couldn't start the build. Unable to run tests.
ohler$
```

If I then `bazel clean` and re-run `bazel test`, it works again – once.

Renaming `_build` to `_work` seemed to make no difference (although I didn't try reproducing that in a clean environment); renaming it to `0_build` makes things work as I expect.  I noticed that `/private/var/tmp/_bazel_ohler/58c6dc864860181b2aa138aee6d4ab43/dir` (where `_build` is being placed) contains directories like `_bin` and `_tmp` that aren't mine, which is why I thought that leading underscores are perhaps special.

The `_build` directory is at the top level of the workspace.

This is on Mac OS 10.10.5 and Bazel 0.1.5.  I'm using a heavily customized version of the Go build rules.

**Comment by rokstrnisa** [comment]
I see this error on Windows. I think Bazel doesn't handle capitalised letters in directories very well, and having some inconsistencies in referencing a directory within BUILD files can lead to this.

    ERROR: Source forest creation failed: Command failed cmd.exe /c mklink /J "C:\tools\msys64\var\tmp\Bazel\$KNt4ZiP\execroot\<dir>" "C:\tools\msys64\home\rok\<dir>".

**Comment by damienmg** [comment]
Can you open a separate bug?

**Comment by rokstrnisa** [comment]
https://github.com/bazelbuild/bazel/issues/1583

**Comment by likanhp-master** [comment]
I am hitting the same issue again, on OSX, with bazel version 0.13.1 and 0.14.1. The directory that causes problem was .builds and renaming it doesn't make any difference.
